fix show function

xinluo2018 opened the pull request:
we found the show function has some problems in ploting  the rgb image, thus we fixed it.

sgillies commented:
@xinluo2018 can you explain what the problems are? Otherwise, I don't know if this work solves them.

xinluo2018 commented:
ok, i found the function show() couldn't plot rgb image if the color interpretation information is missing in the image, then we fixed it that is, when the color information is missing for multi-bands image, the first three bands are regarded the RGB bands, like the qgis software.  
moreover, we also added a new feature of histogram stretching for better image visualization, the previous feature of adjust_band() is included in our new function hist_strech(), therefore we removed the function of adjust_band() in the script.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,33 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
 
             except KeyError:
-                arr = source.read(1, masked=True)
+                arr = source.read((3,2,1), masked=True)
```
why do you assume that bands are store as BGR ?

xinluo2018 replied:
This is from our experience in remote sensing application, that is, if the remote sensing image contains multiple bands, the bands are usually stored by ascending order of the wavelength of bands. The BGR wavelengths are: ～460nm, ~530nm, ~580nm, therefore for a multibands image we set the default index (3,2,1) for the color composition, thus mostly the color composition is the true color composition rather than by setting the default index of (1,2,3). 
moreover, if the image bands are not stored as BGR, our revision also at least show the multibands image as a color composite image,  while the previous script only show the multibands image as a grey image.

xinluo2018 replied:
we add a new feature for image histogram stretching,  this feature is used by an optional parameter for users, which will make the image visualization better. i frequently use it for image visualization in my experiment.

vincentsarago replied:
> This is from our experience in remote sensing application, that is, if the remote sensing image contains multiple bands, the bands are usually stored by ascending order of the wavelength of bands. The BGR wavelengths are: ～460nm, ~530nm, ~580nm, therefore for a multibands image we set the default index (3,2,1) for the color composition, thus mostly the color composition is the true color composition rather than by setting the default index of (1,2,3).

This is a pretty heavy assumption... which after 15years working with Remote Sensing data, I would not make myself ;-) 

>  if the image bands are not stored as BGR, our revision also at least show the multibands image as a color composite image, while the previous script only show the multibands image as a grey image.

Which is why I think adding an `indexes=` options might be a better solution

xinluo2018 replied:
that is also ok, we add a new 'indexes=' options, and the user can specify the bands for color composite when visualize the multibands image. Hope you can check my code again, thanks.

vincentsarago commented:
I feel that this PR complexity the show function and make assumption about the data. Maybe adding an `indexes` option could solve your issues

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,34 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
-
             except KeyError:
-                arr = source.read(1, masked=True)
+                if indexes == None: indexes = (0,1,2)
```
if you want a one-liner I would probably do `indexes = indexes or (0, 1, 2)`

xinluo2018 replied:
thanks for your nice suggestion, the scripts have been modified in the latest commit.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,34 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
             except KeyError:
-                arr = source.read(1, masked=True)
+                if indexes == None: indexes = (0,1,2)
+                indexes = tuple(i+1 for i in indexes)
```
I don't think we need to add `+1` the users should pass valid indexes

xinluo2018 replied:
we also resolved this problem in our latest scripts.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,34 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
         if source.ndim >= 3:
             arr = reshape_as_image(source)
+            if indexes == None: indexes = (0,1,2)
+            if arr.shape[-1]>3: arr = arr[:, :, indexes]
```
here I'll just do something like 

```
arr = reshape_as_image(source[:, :, 0:3])
```

xinluo2018 replied:
we have modified the related scripts in our latest code file, please check it.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,34 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
-
-        arr = reshape_as_image(arr)
+    if strech:
+        if clip_percent: 
+            clip_percent=clip_percent
+        else:
+            clip_percent=2
+        arr = hist_strech(arr, clip_percent)
```
```suggestion
        clip_percent = clip_percent or 2
        arr = hist_strech(arr, clip_percent)
```

xinluo2018 replied:
the scripts have been modified according to your suggestion.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -327,22 +330,18 @@ def show_hist(
         plt.show()
 
 
-def adjust_band(band, kind=None):
```
we shouldn't delete methods, even if they are not used anymore here.

you can add a a deprecation warning if needed

xinluo2018 replied:
the adjust function has been restored in our scripts.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -105,31 +105,34 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
-            arr[ii] = adjust_band(band)
-
-        arr = reshape_as_image(arr)
+    if strech:
```
we don't want to break the API so I feel we should keep `adjust` option and then choose between the current `adjust` method and the clip_percent method 

```python
if adjust:
   if clip_percent:
       arr = hist_strech(arr, clip_percent)
    else:
       # adjust

```

xinluo2018 replied:
we have keep the adjust option in our latest scripts.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -72,21 +74,19 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
     if isinstance(source, tuple):
         arr = source[0].read(source[1])
-
+        print(1111)
```
👀

xinluo2018 replied:
sorry, this is for the code debug, we have deleted it.

vincentsarago commented on rasterio/plot.py:
```diff
@@ -31,8 +31,8 @@ def get_plt():
-def show(source, with_bounds=True, contour=False, contour_label_kws=None,
-         ax=None, title=None, transform=None, adjust=False, **kwargs):
+def show(source, with_bounds=True, contour=False, contour_label_kws=None, indexes=None, 
+         ax=None, title=None, transform=None, clip_percent=None, adjust=True, **kwargs):
```
can we revert to `adjust=False` ?

vincentsarago commented on rasterio/plot.py:
```diff
@@ -75,61 +80,63 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
+                    indexes = (1, 2, 3)
+
+        arr = source.read(indexes, masked=True)
+        arr = reshape_as_image(arr)
```
I've edited this part to reduce the code duplication

vincentsarago commented on rasterio/plot.py:
```diff
@@ -327,9 +334,25 @@ def show_hist(
+    arr_hist = np.nanpercentile(np.array(arr), (clip_percent, 100 - clip_percent))
+    arr = (arr - arr_hist[0]) / (arr_hist[1] - arr_hist[0])
+    arr = np.clip(arr, 0, 1)
+    return arr
```
edited this part for code style issues

vincentsarago commented:
@sgillies I did not 

I'll try tomorrow

xinluo2018 commented:
hi, how about the modified code, could you merge our modifications, that the new function could be used in the new release version of rasterio, thanks.

sgillies commented:
@xinluo2018 I am sorry for the delay and I apologize. I am in favor of this and have several suggestions to make.

sgillies commented on rasterio/plot.py:
```diff
@@ -327,9 +334,25 @@ def show_hist(
         plt.show()
 
 
+def hist_strech(arr, clip_percent):
```
@xinluo2018 will you change this to `contrast_stretch(arr, percent_range=(2.0, 98.0))`? That would make the API more like `skimage.exposure.rescale_intensity()`, which would be nice.

xinluo2018 replied:
we have change the script by following you suggestion.

sgillies commented on rasterio/plot.py:
```diff
@@ -327,9 +334,25 @@ def show_hist(
+    """
+    arr_hist = np.nanpercentile(np.array(arr), (clip_percent, 100 - clip_percent))
+    arr = (arr - arr_hist[0]) / (arr_hist[1] - arr_hist[0])
+    arr = np.clip(arr, 0, 1)
```
Is it necessary to clip the array? Is it not ensured by the math on the line above?

xinluo2018 replied:
For rgb image show, it is not necessary, while for the single-band image show, clip the array is necessary. the test sample can be found at: https://github.com/xinluo2018/rasterio/blob/main/test.ipynb

sgillies commented on rasterio/plot.py:
```diff
@@ -31,8 +31,8 @@ def get_plt():
-def show(source, with_bounds=True, contour=False, contour_label_kws=None,
-         ax=None, title=None, transform=None, adjust=False, **kwargs):
+def show(source, with_bounds=True, contour=False, contour_label_kws=None, indexes=None,
+         ax=None, title=None, transform=None, clip_percent=None, adjust=False, **kwargs):
```
```suggestion
         ax=None, title=None, transform=None, percent_range=None, adjust=False, **kwargs):
```

xinluo2018 replied:
the 'clip_percent' has been changed to 'percent_change'.

sgillies reviewed (changes requested):
@xinluo2018 thank you for the contribution! I have a couple of requests.

sgillies commented on rasterio/plot.py:
```diff
@@ -50,12 +50,17 @@ def show(source, with_bounds=True, contour=False, contour_label_kws=None,
     transform : Affine, optional
         Defines the affine transform if source is an array
+    percent_range: tuple, optional
+        percent_range[0], the minimum values (cumulative percentage) of the histogram for histogram streching, 
```
```suggestion
        percent_range[0], the minimum value (cumulative percentage) of the histogram for histogram streching, 
```

xinluo2018 commented:
happy to be 1 of 158 developers  ^_^
